app/src/main/java/org/mozilla/reference/browser/tabs/LastTabFeature.kt

```diff
@@ -0,0 +1,47 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.reference.browser.tabs
+
+import android.app.Activity
+import mozilla.components.browser.state.selector.findTabOrCustomTabOrSelectedTab
+import mozilla.components.browser.state.state.SessionState
+import mozilla.components.browser.state.state.TabSessionState
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.feature.tabs.TabsUseCases
+import mozilla.components.support.base.feature.LifecycleAwareFeature
+import mozilla.components.support.base.feature.UserInteractionHandler
+
+/**
+ * A feature that removes the tab and selects the parent, if one exists.
+ */
+class LastTabFeature(
+    private val store: BrowserStore,
+    private val tabId: String? = null,
+    private val removeTabUseCase: TabsUseCases.RemoveTabUseCase,
+    private val activity: Activity,
+) : LifecycleAwareFeature, UserInteractionHandler {
+
+    override fun start() = Unit
+    override fun stop() = Unit
+
+    /**
+     * Removes the session if it was opened by an ACTION_VIEW intent
+     * or if it has a parent session and no more history.
+     */
+    override fun onBackPressed(): Boolean {
+        val tab = store.state.findTabOrCustomTabOrSelectedTab(tabId) ?: return false
+
+        return if (tab.source is SessionState.Source.External && !tab.restored) {
+            activity.finish()
+            removeTabUseCase(tab.id)
+            true
+        } else {
+            val hasParent = tab is TabSessionState && tab.parentId != null
+            removeTabUseCase(tab.id, selectParentIfExists = hasParent)
```

Comment:
❓ Mostly for my understanding, I was checking the RemoveTabUsecase ([reducer](https://searchfox.org/mozilla-mobile/source/firefox-android/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TabListReducer.kt#101)), it seems the `tab.parentId != null` check is done there, so do we need to pass `hasParent` as the second param or just `true` would suffice.

Reply:
Good point, we don't have to do the extra check for the `parentId`. When this code was originally written, the use case was not doing this check, so we don't need the refactor to include it as well.

Since we're also not in the `SessionState.Source.External` case, then the `tab is TabSessionState` is also redundant as we're not a `CustomTabSessionState` then.

Reply:
> Since we're also not in the SessionState.Source.External case, then the tab is TabSessionState is also redundant as we're not a CustomTabSessionState then.

`CustomTabSessionState` could still have  source as `SessionState.Source.Internal.CustomTab` right? So just the case that the source is not external is not proof enough that the tab is `TabSessionState` right?

Reply:
Interesting point! If a custom tab is internal then we probably do want to finish the activity too because we would be in a context of a separate activity (external app activity).

The secondary value of the `tab is TabSessionState && tab.parentId != null` check is for the return value to tell our other features that we've handled the back press in this feature.

So maybe the if-statement above should be:

```kotlin
val isExternalOrCustomTab = tab.source is SessionState.Source.External ||
    tab.source is SessionState.Source.Internal.CustomTab

return if (isExternalOrCustomTab && !tab.restored) {
    activity.finish()
    removeTabUseCase(tab.id)
    true
} else {
    val hasParent = tab is TabSessionState && tab.parentId != null
    removeTabUseCase(tab.id, selectParentIfExists = hasParent) // Alternatively, set this to always true.
    // We want to return to home if this session didn't have a parent session to select.
    hasParent
}
```

What do you think?